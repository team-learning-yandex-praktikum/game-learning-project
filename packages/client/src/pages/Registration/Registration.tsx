import { FC } from 'react'
import { Routes } from '@routes/constants'
import { fieldsConfig } from '@utils/validation/fieldsConfig'
import { useAuthentication } from '@utils'
import { omit } from 'lodash'
import { useAppDispatch } from '@store/hooks'
import { registration } from '@store/user'
import { RequestRegistrationData } from '@store/user/types'
import AuthForm from '@components/AuthForm'

const Registration: FC = () => {
    const dispatch = useAppDispatch()
    const onRegistration = useAuthentication(
        async (data: RequestRegistrationData) => {
            await dispatch(registration(data)).unwrap()
        }
    )

    return (
        <AuthForm
            title={'Регистрация'}
            fields={omit(fieldsConfig, [
                'displayName',
                'oldPassword',
                'newPassword',
                'newPasswordRepeat',
            ])}
            SubmitButtonProps={{
                children: 'Сохранить',
            }}
            LinkProps={{
                children: 'Вход',
                to: Routes.Login,
            }}
            onSubmit={onRegistration}
        />
    )
}

export default Registration
