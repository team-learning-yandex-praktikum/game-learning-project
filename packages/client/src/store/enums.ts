export enum LoadStatus {
    idle = 'idle',
    loading = 'loading',
    complete = 'complete',
    failed = 'failed',
}
