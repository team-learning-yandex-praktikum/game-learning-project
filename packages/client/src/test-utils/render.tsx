import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { store } from '@store'
import { ReactNode } from 'react'

export const withProviders = (component: ReactNode) => {
    const testStore = store

    return (
        <BrowserRouter>
            <Provider store={testStore}>{component}</Provider>
        </BrowserRouter>
    )
}
