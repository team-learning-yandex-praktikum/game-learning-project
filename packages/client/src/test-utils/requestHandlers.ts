import { http, HttpResponse } from 'msw'
import { joinURL } from '@utils/url/joinURL'
import { EXTERNAL_SERVER_PATH, SERVER_URL } from '@api/baseApi/constants'
import { UserDTO } from '@api/auth/types'
import { transformData } from '@utils/data'
import { UpdatingProfileDTO } from '@api/user/types'

export const userInfoStub: UserDTO = {
    id: 111,
    login: 'ivan',
    display_name: null,
    avatar: null,
    first_name: 'Иван',
    second_name: 'Иванов',
    email: 'ivan@test.ru',
    phone: '79999999999',
}

const baseServerUrl = joinURL(SERVER_URL, EXTERNAL_SERVER_PATH)
export const requestHandlers = {
    signup: http.post(joinURL(baseServerUrl, 'auth/signup'), () =>
        HttpResponse.text(String(userInfoStub.id))
    ),
    signin: http.post(joinURL(baseServerUrl, 'auth/signin'), () =>
        HttpResponse.text('OK')
    ),
    me: http.get(joinURL(baseServerUrl, 'auth/user'), () =>
        HttpResponse.json(userInfoStub)
    ),
    logout: http.post(joinURL(baseServerUrl, 'auth/logout'), () =>
        HttpResponse.text('OK')
    ),
    updatePassword: http.put(joinURL(baseServerUrl, 'user/password'), () =>
        HttpResponse.text('OK')
    ),
    updateProfile: http.put(
        joinURL(baseServerUrl, 'user/profile'),
        async ({ request }) => {
            const body = (await request.json()) as UpdatingProfileDTO
            return HttpResponse.json(
                transformData.from.dto({
                    ...userInfoStub,
                    ...body,
                })
            )
        }
    ),
}
