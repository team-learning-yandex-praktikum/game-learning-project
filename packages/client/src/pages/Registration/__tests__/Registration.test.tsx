import Registration from '../Registration'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import { withProviders } from '../../../test-utils/render'
import { setupServer } from 'msw/node'
import { store } from '@store'
import { LoadStatus } from '@store/enums'
import {
    requestHandlers,
    userInfoStub,
} from '../../../test-utils/requestHandlers'
import { fieldsConfig } from '@utils'

const fieldLabels = {
    firstName: fieldsConfig.firstName.label,
    secondName: fieldsConfig.secondName.label,
    login: fieldsConfig.login.label,
    password: fieldsConfig.password.label,
    repeatedPassword: fieldsConfig.repeatPassword.label,
    email: fieldsConfig.email.label,
    phone: fieldsConfig.phone.label,
    submit: 'Сохранить',
    link: 'Вход',
}

const server = setupServer(requestHandlers.signup, requestHandlers.me)

const getFields = () => {
    const firstName = screen.getByLabelText(fieldLabels.firstName)
    const secondName = screen.getByLabelText(fieldLabels.secondName)
    const login = screen.getByLabelText(fieldLabels.login)
    const password = screen.getByLabelText(fieldLabels.password)
    const repeatedPassword = screen.getByLabelText(fieldLabels.repeatedPassword)
    const email = screen.getByLabelText(fieldLabels.email)
    const phone = screen.getByLabelText(fieldLabels.phone)

    return {
        firstName,
        secondName,
        login,
        password,
        repeatedPassword,
        email,
        phone,
    }
}

const fillFields = async () => {
    const {
        firstName,
        secondName,
        login,
        password,
        repeatedPassword,
        email,
        phone,
    } = getFields()

    await userEvent.type(firstName, userInfoStub.first_name)
    await userEvent.type(secondName, userInfoStub.second_name)
    await userEvent.type(login, userInfoStub.login)
    await userEvent.type(password, 'SomePass1234')
    await userEvent.type(repeatedPassword, 'SomePass1234')
    await userEvent.type(email, userInfoStub.email)
    await userEvent.type(phone, userInfoStub.phone)
}

describe('Registration', () => {
    beforeAll(() => server.listen())
    afterEach(() => server.resetHandlers())
    afterAll(() => server.close())

    const testedComponent = withProviders(<Registration />)

    test('correctly rendered', async () => {
        const { asFragment } = render(testedComponent)

        expect(asFragment()).toMatchSnapshot()
    })

    test('submit should be disabled', () => {
        render(testedComponent)
        const submitButton = screen.getByText(fieldLabels.submit)
        expect(submitButton).toBeDisabled()
    })

    test('should be disabled without required field', async () => {
        render(testedComponent)

        const login = screen.getByLabelText(fieldLabels.login)
        const submitButton = screen.getByText(fieldLabels.submit)

        await userEvent.type(login, userInfoStub.login)

        expect(submitButton).toBeDisabled()
    })

    test('submit should be enabled with correct values', async () => {
        render(testedComponent)

        const submitButton = screen.getByText(fieldLabels.submit)
        await fillFields()

        expect(submitButton).toBeEnabled()
    })

    test('submit should change state', async () => {
        render(testedComponent)

        const submitButton = screen.getByText(fieldLabels.submit)

        await fillFields()
        await userEvent.click(submitButton)

        const userState = store.getState().user

        expect(userState?.status).toBe(LoadStatus.complete)
        expect(userState?.data.id).toBe(userInfoStub.id)
    })
})
