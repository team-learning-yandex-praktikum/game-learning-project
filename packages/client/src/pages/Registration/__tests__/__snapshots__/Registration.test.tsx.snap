// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration correctly rendered 1`] = `
<DocumentFragment>
  <div>
    <h2
      class=""
    >
      Регистрация
    </h2>
    <form
      id="r:id"
    >
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="firstName"
          >
            Имя
          </label>
          <input
            id="firstName"
            name="firstName"
            placeholder="Иван"
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="secondName"
          >
            Фамилия
          </label>
          <input
            id="secondName"
            name="secondName"
            placeholder="Иванов"
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="email"
          >
            Почта
          </label>
          <input
            id="email"
            name="email"
            placeholder="ivan@test.ru"
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="phone"
          >
            Телефон
          </label>
          <input
            id="phone"
            name="phone"
            placeholder="+7 999 999 99 99"
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="login"
          >
            Логин
          </label>
          <input
            id="login"
            name="login"
            placeholder="ivan_ivanov"
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="password"
          >
            Пароль
          </label>
          <input
            id="password"
            name="password"
            placeholder="qwerty"
            type="password"
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="repeatPassword"
          >
            Пароль (ещё раз)
          </label>
          <input
            id="repeatPassword"
            name="repeatPassword"
            placeholder="qwerty"
            type="password"
          />
        </div>
        <div />
      </div>
    </form>
    <div>
      <button
        class=""
        disabled=""
        form="r:id"
        type="submit"
      >
        Сохранить
      </button>
      <a
        href="/login"
      >
        Вход
      </a>
    </div>
  </div>
</DocumentFragment>
`;
